src/irr/asset/CImageLoaderTGA.cpp

```diff
@@ -112,78 +189,133 @@ bool CImageLoaderTGA::isALoadableFileFormat(io::IReadFile* _file) const
-		imageExtent.depth
-	};
+		imgInfo.format = srcFormat;
+		auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(1u);
+		ICPUImage::SBufferCopy& region = regions->front();
+
+		region.imageSubresource.mipLevel = 0u;
+		region.imageSubresource.baseArrayLayer = 0u;
+		region.imageSubresource.layerCount = 1u;
+		region.bufferOffset = 0u;
+		region.bufferRowLength = asset::IImageAssetHandlerBase::calcPitchInBlocks(imgInfo.extent.width, asset::getTexelOrBlockBytesize(srcFormat));
+		region.bufferImageHeight = 0u;
+		region.imageOffset = { 0u, 0u, 0u };
+		region.imageExtent = imgInfo.extent;
+
+		inputCreationImage = asset::ICPUImage::create(std::move(imgInfo));
+		inputCreationImage->setBufferAndRegions(std::move(texelBuffer), regions);
+
+		bool OpenGlFlip = !flip;
+		if (OpenGlFlip)
+		{
+			auto format = inputCreationImage->getCreationParameters().format;
+			asset::TexelBlockInfo blockInfo(format);
+			core::vector3du32_SIMD trueExtent = blockInfo.convertTexelsToBlocks(inputCreationImage->getRegions().begin()->getTexelStrides());
 
-	auto channels = getFormatChannelCount(destFormat);
-	auto stride = regionBufferRowLenght * getTexelOrBlockBytesize(destFormat);
-	
-	if (flip)
-		out += size.X * size.Y * channels;
-	
-	for (int y = 0; y < size.Y; ++y) {
-		if (flip)
-			out -= stride;
-		
-		const void *src_container[4] = {in, nullptr, nullptr, nullptr};
-		convertColor<srcFormat, destFormat>(src_container, out, stride, size);
-		in += stride;
-		
-		if (!flip)
-			out += stride;
+			auto entry = reinterpret_cast<uint8_t*>(inputCreationImage->getBuffer()->getPointer());
+			auto end = entry + inputCreationImage->getBuffer()->getSize();
+			auto stride = trueExtent.X * getTexelOrBlockBytesize(format);
+
+			for (uint32_t y = 0, yRising = 0; y < trueExtent.Y; y += 2, ++yRising)
+				std::swap_ranges(entry + (yRising * stride), entry + ((yRising + 1) * stride), end - ((yRising + 1) * stride));
+		}
 	}
-}
+
+	if (srcFormat == destFormat)
+		newConvertedImage = inputCreationImage;
+	else
+	{
+		using CONVERSION_FILTER = CConvertFormatImageFilter<srcFormat, destFormat>;
+		CONVERSION_FILTER convertFilter;
+		CONVERSION_FILTER::state_type state;
+		{
+			auto referenceImageParams = inputCreationImage->getCreationParameters();
+			auto referenceBuffer = inputCreationImage->getBuffer();
+			auto referenceRegions = inputCreationImage->getRegions();
+			auto referenceRegion = referenceRegions.begin();
+			const auto newTexelOrBlockByteSize = asset::getTexelOrBlockBytesize(destFormat);
+
+			asset::TexelBlockInfo referenceBlockInfo(referenceImageParams.format);
+			core::vector3du32_SIMD referenceTrueExtent = referenceBlockInfo.convertTexelsToBlocks(referenceRegion->getTexelStrides());
+
+			auto newImageParams = referenceImageParams;
+			auto newCpuBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(referenceTrueExtent.X * referenceTrueExtent.Y * referenceTrueExtent.Z * newTexelOrBlockByteSize);
```

Comment:
you're using the old image fromat (which will be smaller) to determine the size of the backing buffer (`referenceTrueExtent`) for the new image (which will be larger)

Reply:
look at `newTexelOrBlockByteSize`

Reply:
yyyh wait, logic in that case is wrong even though

Reply:
fixed in upcoming commit